### PR TITLE
Adds types for PEC

### LINKED ISSUE #369: Create a type for the decomposition of a noisy operation

### PR DESCRIPTION
**Pull request by rmlarose**
Fixes #369.

Note 1: Forked from PR #373, relevant files are in pec/types. 

Note 2: I will follow up with additional PRs on other types to make reviewing more manageable.

### CODE REVIEW

File: mitiq/pec/types/test_types.py
Comment (andreamari) [inline mitiq/pec/types/test_types.py:47]: Here and for all the tests below, the size of the real matrix should be d^2 x d^2 instead of d x d.

Comment (andreamari) [reply mitiq/pec/types/test_types.py:47]: Maybe you could use: 
`real = np.eye(4)` (this is for sure a valid CPTP operation :-) ) 
or
`real = np.kron([[0.51, 0.49], [0.46, -0.45]], [[0.51, 0.49], [0.46, -0.45]]) `

**Review by andreamari — CHANGES_REQUESTED** [review]
Thanks @rmlarose this looks good and useful!
There is a problem with the choice and the dimension of the real matrix associated with a `NoisyOperation`. See related comments.

**Review by andreamari — APPROVED** [review]
Thanks @rmlarose! LGTM!
Note: there are some conflicts so tests did not run.

**Comment by rmlarose** [comment]
Thanks @andreamari. I'll wait for one more review from @karalekas and/or @nathanshammah before merging.

**Comment by rmlarose** [comment]
Now partially fixes #440 and partially fixes #441 (handling conversions in PEC) via conversions in `NoisyOperation`. These issues will be completely fixed once the new types are used in the functions described in these issues.

**Comment by rmlarose** [comment]
@andreamari please take a look at the new changes when you get a chance.

File: mitiq/pec/types/types.py
Comment (andreamari) [inline mitiq/pec/types/types.py:368]: Is there any advantage in using a `cirq.LinearDict` instead of a standard Python `Dict`?

**Review by andreamari — COMMENTED** [review]
Thanks @rmlarose!  LGTM up to minor comments.
P.s. I just noticed both `NoisyOperation` and `OperationDecomposition` don't have a class docstring. Also many methods don't have a docstring but this could be ok, especially for simple and self-explanatory methods.

File: mitiq/pec/types/types.py
Comment (rmlarose) [reply mitiq/pec/types/types.py:368]: A few - arithmetic (`ldict1 + ldict2`), string formatting

```python
ldict = cirq.LinearDict({cirq.X: 0.5, cirq.Z: -0.5})
print(ldict)
# 0.500*X-0.500*Z
```

etc. Time will tell if they're useful or not, but I suspect they will be.

File: mitiq/pec/types/types.py
Comment (rmlarose) [reply mitiq/pec/types/types.py:369]: Actually - shouldn't we accept any `basis_expansion` since we normalize it to a distribution here?

Comment (andreamari) [reply mitiq/pec/types/types.py]: If the coefficients don't sum up to 1.0 (without taking the absolute value) it means that we are representing some weird non-trace preserving channel. However, I agree that we should probably not enforce it.
If the user, for some strange reason, wants to use some unconventional representation we can allow it.

**Review by andreamari — APPROVED** [review]
LGTM thanks!

File: mitiq/pec/types/types.py
Comment (rmlarose) [reply mitiq/pec/types/types.py]: Sounds good, thanks @andreamari

**Pull request merged**
